### ISSUE TITLE
[engine] Document how to solve a goal.

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#6608]
Solving a goal should be a basic operation of the `Proofview` API.

However, AFAICS, a goal can't be solved using just "safe" operations. Moreover, the invariants of the `Unsafe` operations are not specified.

**Comment by ejgallego** [comment]
In particular I wonder how the goal of "porting the pretyper to the monad" is gonna work given what the current API is. What is the plan? Moving `refine` back to the pretyper?

**Comment by ppedrot** [comment]
No, the pretyper does not need to know about the `refine` tactic. We're merely using the monad to safely thread a state and implement bracktracking on unification, but the pretyper should be unaware of the goal machinery.

**Comment by ejgallego** [comment]
So far, I could deduce that "goal" solving is implemented by manipulating the "`comb`" with `tcl{GET,SET}GOALS`. However, these are in `Unsafe` so they have quite a few invariants, and likely require the use of `tclEVARS`.
